Generate warnings for deprecated algorithms on server startup

edewata opened the pull request:
The PKI server has been modified to generate warnings for deprecated algorithms in the config files and cert profiles when the server is started.

ckelleyRH commented on base/server/scripts/operations:
```diff
@@ -804,10 +804,24 @@ backup_instance_configuration_files()
+    # The result will appear in systemd journal when the server is started.
+
+    FILES=$(find /etc/pki/${pki_instance_id} -type f -not -path "/etc/pki/${pki_instance_id}/alias/*")
+    grep -E "SHA1|SHA-1|SHA_1" $FILES
```
This regex won't pick up `SHA` which is an alias for `SHA-1`. I appreciate that complicates things massively.

edewata replied:
Yeah, if we want to be precise we'll need to handle each configuration parameter individually, which would require a lot of work to implement. I think this static config check with the runtime check you just implemented should be sufficient for now.

edewata replied:
Or how about this?
```
grep -E "SHA$|SHA,|SHAwith|SHA1$|SHA1,|SHA1with|SHA-1$|SHA-1,|SHA_1$|SHA_1," $FILES
````
Do we need to add `SHA-1with` and `SHA_1with` too? Are there other variants that we need to add?

edewata replied:
@jmagne @ladycfu Could you take a look at this? ^^

ckelleyRH replied:
I think you can do better with the regex `^SHA([^2-9]{0,2})$`. Testing on https://www.regexpal.com/
```
SHA (hit)
SHA1 (hit)
SHA-1 (hit)
SHA_1 (hit)
SHA256 (miss)
SHA-256 (miss)
SHA_256 (miss)
```

edewata replied:
Hmm.. that probably wouldn't work on a comma-separated list of algs.

ckelleyRH replied:
It works if you drop the start/end markers so `SHA([^2-9]{0,2})`. I tried at the link before with the test string:
```
policyset.adminCertSet.8.constraint.params.signingAlgsAllowed=SHA256withRSA,SHA512withRSA,SHAwithDSA,SHA1withEC,SHA256withEC,SHA384withRSA,SHA384withEC,SHA512withEC,SHA256withRSA/PSS,SHA384withRSA/PSS,SHA512withRSA/PSS
```
It hits at `SHAwithDSA`, which is right. If you remove that it hits `SHA1withEC`. If you remove that it misses the string.

ckelleyRH replied:
It won't find every instance though, but maybe you could fix it so it does?

edewata replied:
Without the start/end markers it would match `SHA384withRSA`.

ckelleyRH replied:
Oh sorry it should have been `SHA([^2-9]{1,2})`, then it only matches `SHA` and the various permutations of `SHA-1`, and doesn't hit the others. I was very confused for a second there :smile:

jmagne reviewed (commented):
It looks like we are only searching through /etc/pki? Don't we want to search through the profiles directory, which has the various allowed signing algs sections? Or maybe you wanted this limited to only to configs such as CS.cfg, etc.. As for the search string could we just add SHAwith to grab those instances?

edewata commented:
@ckelleyRH @jmagne Thanks for the comments!

I've updated the PR to check the profiles. I also updated the regex. Is it sufficient for now?
It should generate something like this:
```

WARNING: Deprecated algorithm in /etc/pki/pki-tomcat/ca/CS.cfg:378: ca.Policy.rule.SigningAlgRule.algorithms=SHAwithRSA
```

jmagne reviewed (approved):
I actually tried this out by making my own little script and it seems to be working. Both with respect to checking the profiles directory and grabbing the offensive algs. I can defer further to ckelley, if he wants further improvement to the regex.

ckelleyRH reviewed (approved):
> I actually tried this out by making my own little script and it seems to be working. Both with respect to checking the profiles directory and grabbing the offensive algs. I can defer further to ckelley, if he wants further improvement to the regex.

If you have satisfied yourself the regex works with your script then I am satisfied the regex works also :+1: `SHA` is the trickiest one to spot and remove, so I think it is important that we took the time now to make sure we're finding it as it will prevent headaches further down the line.

ckelleyRH commented:
I realised that I had made a typo when presenting my alternate regex, it does work. I have no intention of pushing for its inclusion though, if the existing one in the patch works and is tested that is more than good enough for me :smiley:

edewata commented:
@jmagne @ckelleyRH Thanks! I'll merge it.

The suggested regex was a good one (and I learned something new!), it's very concise for detecting SHA1 variants, but since we will potentially deprecate more algorithms in the future, it might be easier to simply add more matching rules rather than revise the existing regex to match the new set of algorithms. In the future we can move these rules into a file so it'll be easier to manage.

ckelleyRH commented:
It's funny you should mention additional deprecations https://github.com/dogtagpki/jss/pull/828/files :rofl:
